DIPs/DIP1xxx-JS.md

```diff
@@ -0,0 +1,121 @@
+used across DMD, Druntime, and Phobos.
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Acknowledgements](#acknowledgements)
+* [Reviews](#reviews)
+
+## Rationale
+
+There is general disagreement on the best and/or accepted way to remove public
+features. Each deprecation ends up being handled slightly differently depending on
+who's handling the pull requests. Standardizing the process makes sure that
+deprecations are done very publicly and carefully, so as to minimize breakage and
+to provide clear fixes for user code.
+
+## Description
+
+The following lays out the rules to follow to deprecate a part of D.
+
+A symbol or feature must not be marked for removal on a specific date, but rather on a
+specific release. This allows users to easily know if upgrading will break their
+code or not.
+
+Users must be given at least 10 non-patch releases before the deprecated features
+are removed. More releases can be given if the removed code is commonly used.
+There are two cases where the deprecation period is allowed to be shorter:
+
+1. The code or feature is notably dangerous or unsafe, and users need to remove
+it from their code as soon as possible.
+2. The existance of the current code precludes its own fix or the fix of an equally
+important issue.
+
+Shorting the deprecation period should be done with caution to avoid giving D
+an image of instability.
+
+At the time of the pull request for deprecation, all code in Phobos, Druntime,
+and DMD must be updated to remove use of the effected code. Any projects that
+are tested on the Project Tester that are broken should also have their
+maintainers notified.
+
+Both at the time of deprecation and removal, a changelog entry must be made. This
+changelog entry should have a short motivation for the deprecation (or removal)
+and should to describe what steps can be taken by the user to upgrade their codebase.
+
+In order to facilitate on schedule deprecations, a comment of the format
+`@@@DEPRECATED_[version]@@@` should be added to the top of the code to be removed/disabled.
+This comment allows code to be easily searched before every release to
+catch all planned deprecations.
+
+### Public Functions, Types, and Modules
+
+All removals (or changes that make the code `private`/`package`) of public functions,
+types, and modules must be accompanied with a deprecation period.
+
+The symbol(s) must be marked using the `deprecated` keyword with a message containing
+the planned removal period. A pointer to more information should also be added. E.g.
+"See the 2.080 changelog for more details" or "See the function documentation for more
+details". The documentation of the symbol(s) must be updated noting the
+deprecation and removal plan. The documentation should contain information to help
+the users using the symbol(s) transition their code away from the symbol(s).
+
+If the deprecation is occuring because the symbol(s) are being replaced by new
```

Comment:
I don't understand the idea of keeping both old and new code in an un-deprecated state when its known that the old code will be deprecated - what's the reasoning behind this?

It looks like it will just increase the creation of legacy code. For [a trivial] example, if we were to rename std.algorithm.searching's "canFind" to "contains", I would want to know that canFind was deprecated as soon as contains was available so I don't keep writing canFind until the next release, and now have to change even more code. For a mere name substitution, this really doesn't matter, but if I need to do more it would be annoying.

Many people that want their projects to build on multiple compiler versions seem to go for at least three releases, so a single un-deprecated release doesn't solve the problem of supporting multiple versions with the same code; many people will need use version() statements anyway.

Reply:
Another case is supporting e.g. both DMD and LDC. It's not reasonably possible without a window.

Reply:
@jmdavis @CyberShadow Would it be too much overhead to document the old code as deprecated, but not mark it in the source until the next version? This way the compiler is happy and its clear to someone reading the docs that they don't want to use the old function.

Reply:
That's basically how this works now. If a new symbol is replacing an old one, then the new one gets added, and a note gets added to the documentation for the old one that it's going to be deprecated in the next release. Then, in the next release, it's deprecated, and the normal deprecation cycle starts.

If no new symbol is being introduced to replace the symbol being deprecated, then it is deprecated immediately, and the normal deprecation cycle starts.